contracts/Compass.vy

```diff
@@ -225,6 +235,14 @@ def submit_batch(consensus: Consensus, token: address, args: TokenSendArgs, batc
     self.last_batch_id[token] = batch_id
     log BatchSendEvent(token, batch_id, _nonce, _event_id)
 
+@external
+def emit_nodesale_event(buyer: address, paloma: bytes32, node_count: uint256, grain_amount: uint256):
```

Comment:
How do we make sure I don't just call this myself a few times to get free nodes on Paloma?

Reply:
We don't know nodesale contract address before compass-evm deployed. So after fetching event, you need to filter the event with contract address parameter in the NodeSale event. @byte-bandit

Reply:
@byte-bandit we save the node sale contract addresses in paloma, and check them before creating the node licenses, [here](https://github.com/palomachain/paloma/blob/master/x/skyway/keeper/attestation_handler.go#L273).

Reply:
SGTM!